packages/beacon-node/src/chain/regen/regen.ts

```diff
@@ -10,29 +10,34 @@ import {
   stateTransition,
 } from "@lodestar/state-transition";
 import {IForkChoice, ProtoBlock} from "@lodestar/fork-choice";
-import {sleep} from "@lodestar/utils";
+import {Logger, sleep} from "@lodestar/utils";
 import {SLOTS_PER_EPOCH} from "@lodestar/params";
 import {ChainForkConfig} from "@lodestar/config";
 import {Metrics} from "../../metrics/index.js";
 import {IBeaconDb} from "../../db/index.js";
-import {CheckpointStateCache, StateContextCache} from "../stateCache/index.js";
 import {getCheckpointFromState} from "../blocks/utils/checkpoint.js";
 import {ChainEvent, ChainEventEmitter} from "../emitter.js";
+import {CheckpointStateCache, BlockStateCache} from "../stateCache/types.js";
 import {IStateRegeneratorInternal, RegenCaller, StateCloneOpts} from "./interface.js";
 import {RegenError, RegenErrorCode} from "./errors.js";
 
 export type RegenModules = {
   db: IBeaconDb;
   forkChoice: IForkChoice;
-  stateCache: StateContextCache;
+  stateCache: BlockStateCache;
   checkpointStateCache: CheckpointStateCache;
   config: ChainForkConfig;
   emitter: ChainEventEmitter;
+  logger: Logger;
   metrics: Metrics | null;
 };
 
 /**
  * Regenerates states that have already been processed by the fork choice
+ * Since Feb 2024, we support reloading checkpoint state from disk via shouldReload flag. Due to its performance impact
+ * this flag is only set to true in this case:
+ *    - getPreState: this is for block processing, it's important to reload state in unfinality time
+ *    - updateHeadState: rarely happen, but it's important to make sure we always can regen head state
```

Comment:
I checked this PR, and the original PR, it seems like `updateHeadState` is not calling `getState()` with `shouldReload = true`. Is that intentional?

Reply:
no it should call `getState()` with `shouldReload=true`, it's implemented here https://github.com/ChainSafe/lodestar/pull/6359/files#diff-f58ab67de5a502827fc0427a3536a0644e536ddc86fb8f93690e99c499eb7e0cR129

I only want to make mainly `regen` change in this PR to make it easier to review, that'll come next

for the current config, we always go with old state caches so `shouldReload` does not affect at all